pkg/southbound/synchronizer/deviceUpdate.go

```diff
@@ -0,0 +1,158 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package synchronizer
+
+import (
+	"strconv"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/cenkalti/backoff"
+
+	"github.com/onosproject/onos-config/pkg/events"
+
+	topodevice "github.com/onosproject/onos-topo/api/device"
+)
+
+func (sm *SessionManager) updateDevice(id topodevice.ID, connectivity topodevice.ConnectivityState, channel topodevice.ChannelState,
+	service topodevice.ServiceState) error {
+	log.Info("Update device state")
+
+	topoDevice, err := sm.deviceStore.Get(id)
+	st, ok := status.FromError(err)
+
+	// If the device doesn't exist then we should not update its state
+	if ok && err != nil && st.Code() == codes.NotFound {
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
+
+	protocolState, index := containsGnmi(topoDevice.Protocols)
+	if protocolState != nil {
+		topoDevice.Protocols = remove(topoDevice.Protocols, index)
+	} else {
+		protocolState = new(topodevice.ProtocolState)
+	}
+
+	protocolState.Protocol = topodevice.Protocol_GNMI
+	protocolState.ConnectivityState = connectivity
+	protocolState.ChannelState = channel
+	protocolState.ServiceState = service
+	topoDevice.Protocols = append(topoDevice.Protocols, protocolState)
+	mastershipState, err := sm.mastershipStore.GetMastership(topoDevice.ID)
+	if err != nil {
+		return err
+	}
+
+	if topoDevice.Attributes == nil {
+		topoDevice.Attributes = make(map[string]string)
+	}
+
+	topoDevice.Attributes[mastershipTermKey] = strconv.FormatUint(uint64(mastershipState.Term), 10)
+	_, err = sm.deviceStore.Update(topoDevice)
+	if err != nil {
+		log.Errorf("Device %s is not updated %s", id, err.Error())
+		return err
+	}
+	log.Infof("Device %s is updated with states %s, %s, %s", id, connectivity, channel, service)
+	return nil
+}
+
+func containsGnmi(protocols []*topodevice.ProtocolState) (*topodevice.ProtocolState, int) {
+	for i, p := range protocols {
+		if p.Protocol == topodevice.Protocol_GNMI {
+			return p, i
+		}
+	}
+	return nil, -1
+}
+
+func remove(s []*topodevice.ProtocolState, i int) []*topodevice.ProtocolState {
+	s[i] = s[len(s)-1]
+	return s[:len(s)-1]
+}
+
+func (sm *SessionManager) updateDeviceState() error {
+
+	for event := range sm.southboundErrorChan {
+		log.Info("update event received")
+		switch event.EventType() {
+		case events.EventTypeDeviceConnected:
+			log.Info("Device connected")
+			id := topodevice.ID(event.Subject())
+			// TODO: Retry only on write conflicts
+			return backoff.Retry(func() error {
+
+				/* This entire read-modify-write sequence has to be retried until one of two conditions is met:
+				   - The update is successful or
+				   - Upon retry, the node encounters a mastership term greater than its own */
+
+				session := sm.sessions[id]
+				if session != nil {
+					currentTerm, err := session.getCurrentTerm()
```

Comment:
Even though we’re getting the device’s current term from the topology and checking it against the local mastership term, this particular implementation doesn’t guarantee the update cannot be overwritten by an older master. In order to prevent that, the read-modify-write sequence comparing the stored term with the local term and then updating the device _must be atomic_, and in order for it to be atomic it _must operate on the same revision_ of the device. When a device is read from the topology service, the device is returned with metadata that includes a revision number. When a device is updated, the topology service compares the device’s revision number with the stored revision number and rejects the update if the revisions don’t match. If an update is rejected, the client attempts the atomic read-modify-write sequence again.

But the code here is actually reading the device twice: once to retrieve the term, and once to update the device, so the sequence is not atomic. The term can change between the first and second read, and that would allow an older master to overwrite an update from a newer master. For example:
* Node A reads the device and verifies the stored term matches its local term
* Node B is elected master for a later term and updates the device state
* Node A proceeds to read the device again and updates it _without checking the term_, overwriting the update from new master node B

Both the condition checking the device’s stored term and the write updating the device state must occur on the same device object returned from the topology service.